mmv1/third_party/terraform/data_sources/data_source_google_firebase_web_app_config.go.erb

```diff
@@ -3,133 +3,197 @@ package google
+				Description:         "The default Cloud Storage for Firebase storage bucket name.",
+				MarkdownDescription: "The default Cloud Storage for Firebase storage bucket name.",
+				Computed:            true,
 			},
-			"messaging_sender_id": {
-				Type:        schema.TypeString,
-				Computed:    true,
-				Description: `The sender ID for use with Firebase Cloud Messaging.`,
+
+			"project": schema.StringAttribute{
+				Description:         "The project id of the Firebase web App.",
+				MarkdownDescription: "The project id of the Firebase web App.",
+				Optional:            true,
 			},
-			"storage_bucket": {
-				Type:        schema.TypeString,
-				Computed:    true,
-				Description: `The default Cloud Storage for Firebase storage bucket name.`,
+
+			"id": schema.StringAttribute{
+				Description:         "Firebase Web App Config identifier",
+				MarkdownDescription: "Firebase Web App Config identifier",
+				Computed:            true,
 			},
 		},
 	}
-
 }
 
-func dataSourceGoogleFirebaseWebappConfigRead(d *schema.ResourceData, meta interface{}) error {
-	config := meta.(*Config)
-	userAgent, err := generateUserAgentString(d, config.UserAgent)
-	if err != nil {
-		return err
+func (d *GoogleFirebaseWebAppConfigDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
 	}
 
-	id := d.Get("web_app_id").(string)
-
-	project, err := getProject(d, config)
-	if err != nil {
-		return err
+	p, ok := req.ProviderData.(*frameworkProvider)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *frameworkProvider, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
 	}
 
-	url, err := replaceVars(d, config, "{{FirebaseBasePath}}projects/{{project}}/webApps/{{web_app_id}}/config")
-	if err != nil {
-		return err
+	d.client = p.NewFirebaseClient(p.userAgent, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
 	}
+	d.project = p.project
+}
 
-	res, err := SendRequest(config, "GET", project, url, userAgent, nil)
-	if err != nil {
-		return handleNotFoundError(err, d, fmt.Sprintf("FirebaseWebApp config %q", d.Id()))
-	}
+func (d *GoogleFirebaseWebAppConfigDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data GoogleFirebaseWebAppConfigModel
+	var metaData *ProviderMetaModel
 
-	err = d.Set("api_key", res["apiKey"])
-	if err != nil {
-		return err
-	}
-	err = d.Set("auth_domain", res["authDomain"])
-	if err != nil {
-		return err
-	}
-	err = d.Set("database_url", res["databaseURL"])
-	if err != nil {
-		return err
-	}
-	err = d.Set("location_id", res["locationId"])
-	if err != nil {
-		return err
+	// Read Provider meta into the meta model
+	resp.Diagnostics.Append(req.ProviderMeta.Get(ctx, &metaData)...)
```

Comment:
thinking out loud, not necessarily actionable: I've been finding the `resp.Diagnostics.Append(doStuff())` pattern hard to read- I keep focusing on the part where we're writing to diagnostics, rather than the action we're taking . Did you find it grew on you? That was the case with errors in Go for me, initially, so it may just be that this looks different than I'm used to.

I've been wondering if a pattern like the following would be more clear to folks off the bat:

```
diag := req.ProviderMeta.Get(ctx, &metaData)
resp.Diagnostics.Append(diag...)
if resp.Diagnostics.HasError() {
	return
}
```

Even better (imo) would be if there was a function in the framework like "IsEmpty()" or "HasAny()", allowing the control flow to be more clear:

```
diag := req.ProviderMeta.Get(ctx, &metaData)
if !diag.IsEmpty() {
	resp.Diagnostics.Append(diag...)
	if resp.Diagnostics.HasError() {
		return
	}
}
```

Reply:
I don't have strong enough feelings to say we need to make this change, I'm just overthinking this stuff now to try and preempt needing to change it later! 

> As for the second example, I don't believe they have any of those methods currently available

Yeah, they're not. They'd be trivial to PR in, I think!